docs/available-filters/envelope-encryption.adoc

```diff
@@ -0,0 +1,155 @@
+<4> The key selector name.
+<5> Object providing configuration understood by key selector.
+
+==== KMS configuration
+
+For the KMS configuration:
+
+===== HashiCorp Vault
+
+For HashiCorp Vault, the KMS configuration looks like this:
+
+[source, yaml]
+----
+kms: VaultKmsService                                          # <1>
+kmsConfig:
+  vaultUrl: http://vault.vault.svc.cluster.local:8200         # <2>
+  vaultToken: myroottoken                                     # <3>
+----
+<1> The name of the KMS provider. This must be `VaultKmsService`.
+<2> The Vault URL
+<3> The Vault Token
+
+==== Key Encryption Key (KEK) selector configuration
+
+The role of the Key Encryption Key (KEK) selector to to map from the
+
+===== Template KEK Selector
+
+The `TemplateKekSelector` maps from topic name to key name.  The template understand the token
+`$\{topicName}` which is replaced by the name of the topic.  It can be used to build key names
+that include the topic name being encrypted.  Use the `$\{topicName}` is optional. It is possible
+to pass a literal string.  This will result in all topics being encrypted using the same key.
+
+[source, yaml]
+----
+selector: TemplateKekSelector                                 # <1>
+selectorConfig:
+  template: "key_${topicName}"                                # <2>
+----
+<1> The name of the KEK selector. This must be `TemplateKekSelector`.
+<2> Template used to map for the topic name to the name of the key within the KMS.
+
+=== Establishing the keys in the KMS
+
+Use the management interface of the KMS to create the encryption keys.  The names(or aliases) of the encryption keys
+must match the naming conventions established within the configuration of the KEK selector. If this is not done,
+the filter will be unable to find the encryption key, and it won't be able to encrypt the produce requests.
+
+For example, if using the `TemplateKekSelector` with the template `foo_$\{topicName}`, create a key for every topic that
+is to be encrypted with the key name matching the topic name, prefixed by the string `foo`.
+
+CAUTION: If the filter is unable to find a key for a given topic, the filter will pass through the records of the
+produce request without encrypting the records.
+
+==== HashiCorp Vault
+
+Use either the HashiCorp UI or CLI to create AES-256 symmetric keys following your key naming convention.
+
+TIP: It is recommended to apply a key rotation policy to key as it is created.
+
+=== Verifying that encryption is occurring
+
+To verify that records sent to topics are indeed being encrypted, use `kafka-console-consumer` to consume the
+records directly from the target Kafka Cluster.  Verify that encrypted text is seen rather than whatever plain text
+that was sent by producer.  The record values seen will look something like this:
+
+```
+tradesvault:v1:+EfJ977UG1XkjI9yh7vxpgN2E1DKaIkDuxE+eCprVTKr+sskFuChcTe/KpR/c8ZDyP76W3itExmEzLOl����x)�Ũ�z�:S�������tБ��v���
+```
+
```

Comment:
The alternative verification is to look at the metrics.  I thought this was path of least resistence.

Reply:
This is fine. It's more convincing proof, for a start.